info_encoding.md

```diff
@@ -0,0 +1,143 @@
+# Encoding of the `info` Field in Derivations
+
+The key derivation should include all information of the context
+of the derivation.
+This includes information that is valid only for this particular message,
+as well as the intended use of the key.
+
+Once we have agreed on a suitable selection of values to include, we
+also must encode them in such a way that it can not collide with
+the honest encoding of a different set of values.
+Obviously, an adversary can perform a different encoding, but this
+does not matter, as we assume they can not trick the user into performing
+off-standard derivations or use an off-standard encoding for their derivations.
+Implementations must make sure this assumption holds.
+
+## Notation
+
+In this document, we write lists as comma seperated value, enclosed
+in parentheses. Therefore a list of values 1, 2 and 3 would be written
+as `(1, 2, 3)`.
+Binary concatenation is `||`.
+
+## Requirements
+
+We require the format to support lists of buffers.
+This is enough for most simple operations, and can be extended to
+key-value stores.
+
+To get a simpler encoding, we do not require schemalessness.
+Therefore, derivation steps need to define a schema for the encoding
+of the info field.
+For the core derivations, this is given.
+Custom applications need to provide this as well.
+Note that schemalessness would not help in this context, because 
+all users that are derive a key are on the same page what that key
+is supposed to be used for.
+If they are not, the adversary got them to execute their code already,
+which means that particular user is in trouble anyways (REPHRASE).
+
+We require the format to be able to handle buffers that are less than 2<sup>16</sup>B in size.
+
+At this point, we do not require to be able to encode nested data.
+However, we will later propose two options for backwards-compatible
+changes to this proposal that support that.
+
+## Format
+
+The proposed format is called _shallow length prefixed (SLP)_ encoding.
+Shallow, because it does not specify nesting.
+Length-prefixed, because each element is prefixed with its length.
+
+Specifically, the encoding is the element-wise concatenation of the the
+length of the elements and the element.
+The length is encoded as big endian uint16. (REALLY BIG ENDIAN? PRO/CON?)
```

Comment:
I know we used  little-endian encoding for the offset in the header: 
```js 
header.writeUInt16LE(offset, 0)
```

I don't know the pros and cons of each.
Could this be an example of "lets use a consistent pattern to make implementing the spec easier"?

Reply:
If we'd try to sort the data by bytes the endianness would be important, but yes, I think in this case consistency is key. Maybe we should define a byte order somewhere and refer to that, so we can make up our minds later and enforce consistency.

Reply:
Yep, but there's only 2 instances at the moment. Probably too early to abstract

Reply:
TOO LATE!